File: tests/e2e/compare_and_branch.py
# RUN: uv run xdsl-exo -o %t.mlir %s
# RUN: mlir-opt --convert-arith-to-llvm --convert-cf-to-llvm --convert-func-to-llvm --reconcile-unrealized-casts %t.mlir | mlir-translate --mlir-to-llvmir | llc -filetype=obj -o %t.o
# RUN: clang %t.o %S/compare_and_branch.c -o %t.bin -lm
# RUN: %t.bin | filecheck %s

from __future__ import annotations

from exo import *


# CHECK: 10.0
# CHECK: 20.0
@proc
def compare_and_branch(out: f32[1] @ DRAM, a: index, b: index):
    assert a >= 0
    assert b >= 0
    if a < b:
        out[0] = 10.0
    else:
        out[0] = 20.0

File: tests/e2e/compare_and_branch.c
#include <stdio.h>

extern void compare_and_branch(float *out, long a, long b);

int main() {
    float out = 0.0f;

    compare_and_branch(&out, 3, 7);
    printf("%.1f\n", out);

    compare_and_branch(&out, 7, 3);
    printf("%.1f\n", out);

    return 0;
}
